Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -1332,6 +1332,66 @@ The following approach can be used in order to patch the JavaScript file:
+
+A well-known use case of this technique is loading the Frida gadget to an application, specially while working on a non-rooted device (this is what [`objection patchapk`](https://github.com/sensepost/objection/wiki/Patching-Android-Applications "Patching Android Applications") basically does).
+
+###### Patching Application's Native Library
+
+Many Android applications use native code in addition to Java code for various performance and security reasons. The native code is present in the form of ELF shared libraries. An ELF executable includes a list of shared libraries (dependencies) that are linked to the executable for it to function optimally. This list can be modified to insert an additional library to be injected into the process.
+
+Modifying the ELF file structure manually to inject a library can be cumbersome and prone to errors. However, this task can be performed with relative ease using [LIEF](https://lief.quarkslab.com/ "LIEF") (Library to Instrument Executable Formats). Using it requires only a few lines of Python code as shown below:
+
+```python
+import lief
+
+libnative = lief.parse("libnative.so")
+libnative.add_library("libinject.so") # Injection!
+libnative.write("libnative.so")
+```
+
+In the above example, libinject.so library is injected as a dependency of a native library (libnative.so), which the application already loads by default. Frida gadget can be injected into an application using this approach as explained in detail in [LIEF's documentation](https://lief.quarkslab.com/doc/latest/tutorials/09_frida_lief.html "LIEF's documentation - How to use frida on a non-rooted device"). As in the previous section, it is important to remember adding the library to the respective architecture `lib` folder in the APK and finally re-signing the application.
+
+###### LD_PRELOAD
+
+Above we looked into techniques which require some kind of modification of the application's code. A library can also be injected into a process using functionalities offered by the loader of the operating system. On Android, which is a Linux based OS, you can load an additional library by setting the `LD_PRELOAD` environment variable.
+
+The symbols loaded from the library passed using LD_PRELOAD always gets the precedence, i.e they are searched first by the loader while resolving the symbols. This feature is often used to modify behaviour of libc functions - fopen(), read(), write(), strcmp() etc, specially in case of obfuscated programs, where information is deliberately obfuscated to prevent static analysis, for instance, file path passed to fopen() or strings passed to strcmp().
+
+On Android, setting `LD_PRELOAD` is slightly different than compared to other Linux distributions: using the `setprop` command. Below you can see an example for an application with package name `com.foo.bar` (note the additional `wrap.` prefix):
```

Comment:
We can maybe add that this is because every application process is forked from zygote, so you would have to use LD_PRELOAD on the zygote process, which won't work of course. As a circumvention, android added the setprop functionality.